File: backend/profiles/schemas/privacy.py
"""Privacy-related Pydantic schemas for data overview, export, and account deletion."""

from __future__ import annotations

from typing import Any

from ninja import Schema
from pydantic import field_validator


# --- Data Overview Sub-Schemas ---


class CategorySchema(Schema):
    """Generic category with count and items list."""

    count: int
    items: list[dict[str, Any]]


class AnalyticsDataSchema(Schema):
    """Summary of analytics data (no individual records)."""

    view_count: int
    search_count: int


class DataOverviewSchema(Schema):
    """Complete overview of all personal data stored for a user."""

    profile: dict[str, Any]
    preferences: dict[str, Any] | None = None
    groups: CategorySchema
    persons: CategorySchema
    events: CategorySchema
    content: CategorySchema
    comments: CategorySchema
    interactions: CategorySchema
    planning: CategorySchema
    packing_lists: CategorySchema
    shopping_lists: CategorySchema
    analytics: AnalyticsDataSchema


# --- Account Deletion ---


class DeleteAccountRequestSchema(Schema):
    """Request to delete (anonymize) user account."""

    password: str | None = None
    confirmation: str

    @field_validator("confirmation")
    @classmethod
    def validate_confirmation(cls, v: str) -> str:
        if v != "KONTO LÖSCHEN":
            raise ValueError("Bitte bestätige die Löschung mit 'KONTO LÖSCHEN'")
        return v

File: frontend/src/schemas/privacy.ts
/**
 * Zod schemas for Privacy API.
 * MUST stay in sync with backend/profiles/schemas/privacy.py
 */
import { z } from 'zod';

// --- Data Overview Schemas ---

export const CategorySchema = z.object({
  count: z.number(),
  items: z.array(z.record(z.string(), z.unknown())),
});
export type Category = z.infer<typeof CategorySchema>;

export const AnalyticsDataSchema = z.object({
  view_count: z.number(),
  search_count: z.number(),
});
export type AnalyticsData = z.infer<typeof AnalyticsDataSchema>;

export const DataOverviewSchema = z.object({
  profile: z.record(z.string(), z.unknown()),
  preferences: z.record(z.string(), z.unknown()).nullable().optional(),
  groups: CategorySchema,
  persons: CategorySchema,
  events: CategorySchema,
  content: CategorySchema,
  comments: CategorySchema,
  interactions: CategorySchema,
  planning: CategorySchema,
  packing_lists: CategorySchema,
  analytics: AnalyticsDataSchema,
});
export type DataOverview = z.infer<typeof DataOverviewSchema>;

// --- Account Deletion Schemas ---

export const DeleteAccountRequestSchema = z.object({
  password: z.string().nullable(),
  confirmation: z.literal('KONTO LÖSCHEN'),
});
export type DeleteAccountRequest = z.infer<typeof DeleteAccountRequestSchema>;
